### PR TITLE
Task/backend20 add mail service

### PR DESCRIPTION
**Pull request by shirokuronemuri**
Beside the mail service, I've also created couple simple mail templates for it which can be changed or added in future.

### CODE REVIEW

**Comment by shirokuronemuri** [comment]
[Add send mail service](https://trello.com/c/fjFIr1pz/20-add-send-mail-service)

**Comment by shirokuronemuri** [comment]
something's very wrong 😢
first I was trying to resolve the conflict, then it just removed all my installed dependencies for some reason
trying to revert last commits now

**Comment by shirokuronemuri** [comment]
finally got it to work

File: backend/src/core/mail/application/mail-service.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import { CONFIG } from '~/configuration/config';
+import { injectable } from 'inversify';
+import Email from 'email-templates';
+import { MailType } from '~/shared/enums/enums';
+import { MailProps } from '~/shared/types/types';
+
+@injectable()
+export class MailService {
+  async sendEmail(receiver: string, mailType: MailType, props: MailProps): Promise<void> {
+    const transport = await createMailTransport();
```
Comment (VitaliiBedletskyi) [inline backend/src/core/mail/application/mail-service.ts]: Mail transport should be initialized one time during creating di container (see initialization PrismaClient in backend/src/secondary-adapters/postgres/postgres-container-module.ts)

File: backend/src/configuration/mail-transport.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { CONFIG } from './config';
```
Comment (VitaliiBedletskyi) [inline backend/src/configuration/mail-transport.ts]: Better to move it to backend/src/secondary-adapters/mail/mail-transport.ts

File: backend/src/core/mail/application/mail-service.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import { createMailTransport } from '~/configuration/mail-transport';
+import { CONFIG } from '~/configuration/config';
+import { injectable } from 'inversify';
+import Email from 'email-templates';
+import { MailType } from '~/shared/enums/enums';
+import { MailProps } from '~/shared/types/types';
+
+@injectable()
+export class MailService {
```
Comment (VitaliiBedletskyi) [inline backend/src/core/mail/application/mail-service.ts]: It should be the `MailAdapter` in `src/secondary-adapters/mail/` that implements port (Interface) MailAdapterInterface. Create `MailAdapterInterface` in `backend/src/core/common`. As a result, all our services from the application level will depend only on `MailAdapterInterface` and don't know anything about realization mail sending

File: backend/src/core/mail/mail-container-module.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { ContainerModule, interfaces } from 'inversify';
+import { MailService } from './application/mail-service';
+import { CONTAINER_TYPES } from '~/shared/types/types';
+
+const mailContainerModule = new ContainerModule((bind: interfaces.Bind) => {
+  bind<MailService>(CONTAINER_TYPES.MailService).to(MailService);
```
Comment (VitaliiBedletskyi) [inline backend/src/core/mail/mail-container-module.ts]: Should bind `MailAdapterInterface` to `MailAdapter` which I wrote above...

**Review by VitaliiBedletskyi — CHANGES_REQUESTED** [review]

**Comment by shirokuronemuri** [comment]
I created separate mail repository like the user one, and moved mail transport to secondary adapter. to be honest I'm not 100% sure if I did the binding right since I haven't done this before and I don't have any proper ways to test it out at the moment.

**Comment by shirokuronemuri** [comment]
gonna create test route to ensure that the service works first, I'll drop commits tomorrow

File: backend/src/core/mail/application/mail-service.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { inject, injectable } from 'inversify';
+import { MailType } from '~/shared/enums/enums';
+import { CONTAINER_TYPES, MailProps } from '~/shared/types/types';
+import { MailRepository } from '~/core/mail/port/mail-repository';
+
+@injectable()
+export class MailService {
```
Comment (VitaliiBedletskyi) [inline backend/src/core/mail/application/mail-service.ts]: As I understood to send mail you don't require to use MailService, just enough MailRepositoryAdapter, or is it some example? If it usage example please add a comment before the class declaration.

File: backend/src/secondary-adapters/mail/mail-transport.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import nodemailer, { Transporter } from 'nodemailer';
+import { google } from 'googleapis';
+import SMTPConnection from 'nodemailer/lib/smtp-connection';
+import { AsyncContainerModule, interfaces } from 'inversify';
+import { MailRepository } from '~/core/mail/port/mail-repository';
+import { CONTAINER_TYPES } from '~/shared/types/container-type-keys';
+import { MailRepositoryAdapter } from './mail-repository-adapter';
+import SMTPTransport from 'nodemailer/lib/smtp-transport';
+
+const mailTransport = new AsyncContainerModule(async (bind: interfaces.Bind) => {
```
Comment (VitaliiBedletskyi) [inline backend/src/secondary-adapters/mail/mail-transport.ts]: Create separate mail-container-module.ts and move AsyncContainerModule to him.

File: backend/src/secondary-adapters/mail/mail-transport.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  const auth: SMTPConnection.AuthenticationTypeOAuth2 = {
+    type: 'OAuth2',
+    user: CONFIG.MAIL_SERVICE.ADDRESS,
+    clientId: CONFIG.MAIL_SERVICE.CLIENT_ID,
+    clientSecret: CONFIG.MAIL_SERVICE.CLIENT_SECRET,
+    refreshToken: CONFIG.MAIL_SERVICE.REFRESH_TOKEN,
+    accessToken,
+  };
+
+  const transport = nodemailer.createTransport({
```
Comment (VitaliiBedletskyi) [inline backend/src/secondary-adapters/mail/mail-transport.ts]: Leave the logic to initialize mail transport in this file

**Review by VitaliiBedletskyi — CHANGES_REQUESTED** [review]

**Comment by VitaliiBedletskyi** [comment]
Hi, could you please fix remarks and resolve merge conflicts?

**Review by VitaliiBedletskyi — APPROVED** [review]

**Pull request merged**